### PR TITLE
StudentList implementation v1.2

### PR DESCRIPTION
**Pull request by Valelaaa**

### CODE REVIEW

File: src/main/java/com/endava/internship/collections/StudentList.java [hunk 1]
```diff
@@ -1,145 +1,547 @@
...
+        students = new Student[capacity];
+    }
+
+    /**
+     * Constructs an empty list with the specified initial capacity.
+     * @param initialCapacity - sets primary capacity to the given integer
+     * @throws IllegalArgumentException - if given integer (initialCapacity)
+     * is less than 0
+     */
+    public StudentList(int initialCapacity){
```
[the remark this answers is not in the document]
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/StudentList.java]: Done

File: src/main/java/com/endava/internship/collections/StudentList.java [hunk 2]
```diff
@@ -1,145 +1,547 @@
...
+    }
 
+    /**
+     * Constructs a list containing the elements of the specified collection,
+     * in the order they are returned by the collection's iterator.
+     * @param c - the collection whose elements are to be placed into this list
+     * @throws NullPointerException if the specified collection is null
+     */
+    public StudentList(Collection<? extends Student> c){
+        if (Objects.isNull(c))
```
[the remark this answers is not in the document]
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/StudentList.java]: Done

File: src/main/java/com/endava/internship/collections/StudentList.java [hunk 3]
```diff
@@ -1,145 +1,547 @@
 package com.endava.internship.collections;
 
+
+import java.util.Arrays;
```
[the remark this answers is not in the document]
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/StudentList.java:3]: Done)

File: src/main/java/com/endava/internship/collections/StudentList.java [hunk 4]
```diff
@@ -1,145 +1,547 @@
...
+     * @throws IllegalArgumentException - if given integer (initialCapacity)
+     * is less than 0
+     */
+    public StudentList(int initialCapacity){
+        if (initialCapacity >= 0) {
+            capacity = initialCapacity;
+            students = new Student[capacity];
+        }
+        else {
+            throw new IllegalArgumentException("Illegal Capacity: " + initialCapacity);
```
[the remark this answers is not in the document]
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/StudentList.java]: Done)

File: src/main/java/com/endava/internship/collections/StudentList.java [hunk 5]
```diff
@@ -1,145 +1,582 @@
...
 import java.util.Collection;
 import java.util.Iterator;
 import java.util.List;
 import java.util.ListIterator;
+import java.util.NoSuchElementException;
 
+import static java.util.Objects.nonNull;
+import static java.util.Objects.isNull;
+
+class ExceptionMessages {
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/collections/StudentList.java:13]: why not in a separate file?

File: src/main/java/com/endava/internship/collections/StudentList.java [hunk 6]
```diff
@@ -1,145 +1,582 @@
...
+     * @param initialCapacity - sets primary capacity to the given integer
+     * @throws IllegalArgumentException - if given integer (initialCapacity)
+     *                                  is less than 0
+     */
+    public StudentList(final int initialCapacity) {
+        if (initialCapacity >= 0) {
+            capacity = initialCapacity;
+            students = new Student[capacity];
+        } else {
+            throw new IllegalArgumentException(String.format(ExceptionMessages.CONSTRUCTOR_ILLEGAL_ARGUMENT_EXCEPTION_MESSAGE, getClass()));
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/collections/StudentList.java:79]: static import, please, for all exceptions
`format(CONSTRUCTOR_ILLEGAL_ARGUMENT_EXCEPTION_MESSAGE, getClass())`

File: src/main/java/com/endava/internship/collections/Student.java [hunk 7]
```diff
@@ -1,32 +1,57 @@
...
-{
-    private String name;
-    private LocalDate dateOfBirth;
+public class Student {
+    private final String name;
+    private final LocalDate dateOfBirth;
     private String details;
 
-    public Student(String name, LocalDate dateOfBirth, String details) {
+    public Student(final String name, final LocalDate dateOfBirth, String details) {
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/collections/Student.java:14]: `(final String name, final LocalDate dateOfBirth, final String details)`

File: src/main/java/com/endava/internship/collections/Student.java [hunk 8]
```diff
@@ -1,32 +1,57 @@
...
 
 /**
  * The class that defines the element that will be contained by your collection
  */
-public class Student //TODO consider implementing any interfaces necessary for your collection
-{
-    private String name;
-    private LocalDate dateOfBirth;
+public class Student {
+    private final String name;
```
Comment (NataliaIovva) [inline src/main/java/com/endava/internship/collections/Student.java:10]: Is final really need here?

File: src/main/java/com/endava/internship/collections/StudentList.java (hunk 5 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/StudentList.java]: Just for convenience. Already transferred to a separate file

File: src/main/java/com/endava/internship/collections/StudentList.java (hunk 6 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/StudentList.java]: done

File: src/main/java/com/endava/internship/collections/Student.java (hunk 7 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/Student.java]: forgot about details. Done

File: src/main/java/com/endava/internship/collections/Student.java (hunk 8 above)
Comment (Valelaaa) [reply src/main/java/com/endava/internship/collections/Student.java]: I thought that such fields should be immutable. Just removed modificator

**Pull request merged**